File: js/mandelbrot.js
class Complex {
    constructor(x, y) {
        this.x = x;
        this.y = y;
    }

    abs() {
        return Math.sqrt(this.x * this.x + this.y * this.y);
    }

    add(other) {
        this.x += other.x;
        this.y += other.y;
    }

    mul(other) {
        let newX = this.x * other.x - this.y * other.y;
        let newY = this.x * other.y + this.y * other.x;
        this.x = newX;
        this.y = newY;
    }
}

// z(n+1) = z(n)^2 + c
function iterate_mandelbrot(c, maxIters) {
    let z = new Complex(0, 0);
    for (let i = 0; i < maxIters; i++) {
        if (z.abs() >= 2) {
            return i;
        }
        z.mul(z);
        z.add(c);
    }
    return maxIters;
}

let start = Date.now();
let x0 = -2.5, x1 = 1, y0 = -1, y1 = 1;
let cols = 72, rows = 24;
let maxIters = 1000000;

for (let row = 0; row < rows; row++) {
    let y = (row / rows) * (y1 - y0) + y0;
    let str = '';
    for (let col = 0; col < cols; col++) {
        let x = (col / cols) * (x1 - x0) + x0;
        let c = new Complex(x, y);
        let iters = iterate_mandelbrot(c, maxIters);
        if (iters == 0) {
            str += '.';
        } else if (iters == 1) {
            str += '%';
        } else if (iters == 2) {
            str += '@';
        } else if (iters == maxIters) {
            str += ' ';
        } else {
            str += '#';
        }
    }
    console.log(str);
}

let end = Date.now() - start;
console.log(end, 'milliseconds runtime');

File: python/mandelbrot.py
import math
import time

class Complex:
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def abs(self):
        return math.sqrt(self.x * self.x + self.y * self.y)

    def add(self, other):
        self.x += other.x
        self.y += other.y

    def mul(self, other):
        newX = self.x * other.x - self.y * other.y
        newY = self.x * other.y + self.y * other.x
        self.x = newX
        self.y = newY


# z(n+1) = z(n)^2 + c
def iterate_mandelbrot(c, maxIters):
    z = Complex(0, 0)
    for i in range(0, maxIters):
        if z.abs() >= 2:
            return i
        z.mul(z)
        z.add(c)
    return maxIters


start = time.monotonic()
x0 = -2.5; x1 = 1; y0 = -1; y1 = 1
cols = 72; rows = 24
maxIters = 1000000

for row in range(0, rows):
    y = (float(row) / float(rows)) * (y1 - y0) + y0
    str = ''
    for col in range(0, cols):
        x = (float(col) / float(cols)) * (x1 - x0) + x0
        c = Complex(x, y)
        iters = iterate_mandelbrot(c, maxIters)
        if iters == 0:
            str += '.'
        elif iters == 1:
            str += '%'
        elif iters == 2:
            str += '@'
        elif iters == maxIters:
            str += ' '
        else:
            str += '#'
    print(str)

end = time.monotonic() - start
print(end * 1000.0, 'milliseconds runtime')
